tools/iree-e2e-matmul-test.c

```diff
@@ -276,34 +341,31 @@ static iree_status_t get_matmul_sizes(
   return iree_ok_status();
 }
 
-// Helper for reference_matmul_element. f32 case.
-static void reference_matmul_element_f32(
+// Helper for reference_matmul_element. i32*i32->i32 case.
+// This reference implementation takes data as void* and cast those to int
+// and runs a reference matmul on casted int. This reference matmul only
+// works when we fill the buffers in small integers.
+static void reference_matmul_element_generic(
     iree_hal_dim_t m_size, iree_hal_dim_t k_size, iree_hal_dim_t n_size,
     iree_hal_element_type_t lhs_type, iree_hal_element_type_t rhs_type,
-    float* lhs_data, float* rhs_data, float* acc_data, float* result_data,
-    iree_hal_dim_t m, iree_hal_dim_t n) {
-  float acc = acc_data[n + m * n_size];
-  for (iree_hal_dim_t k = 0; k < k_size; ++k) {
-    float lhs_value = lhs_data[k + m * k_size];
-    float rhs_value = rhs_data[n + k * n_size];
-    acc += lhs_value * rhs_value;
-  }
-  result_data[n + m * n_size] = acc;
-}
+    iree_hal_element_type_t acc_type, void* lhs_data, void* rhs_data,
+    void* acc_data, void* result_data, iree_hal_dim_t m, iree_hal_dim_t n) {
+  iree_vm_value_t acc_value =
+      read_matrix_element(m_size, n_size, acc_type, acc_data, m, n);
+  int32_t lhs_i32 = {0}, rhs_i32 = {0}, acc_i32 = {0};
+  read_element_as_int(acc_value, &acc_i32);
 
-// Helper for reference_matmul_element. i8*i8->i32 case.
-static void reference_matmul_element_i8_i8_i32(
-    iree_hal_dim_t m_size, iree_hal_dim_t k_size, iree_hal_dim_t n_size,
-    iree_hal_element_type_t lhs_type, iree_hal_element_type_t rhs_type,
-    int8_t* lhs_data, int8_t* rhs_data, int32_t* acc_data, int32_t* result_data,
-    iree_hal_dim_t m, iree_hal_dim_t n) {
-  int32_t acc = acc_data[n + m * n_size];
   for (iree_hal_dim_t k = 0; k < k_size; ++k) {
-    int8_t lhs_value = lhs_data[k + m * k_size];
-    int8_t rhs_value = rhs_data[n + k * n_size];
-    acc += ((int32_t)lhs_value) * ((int32_t)rhs_value);
+    iree_vm_value_t lhs_value =
```

Comment:
have you profiled this? the e2e matmul tests is usually the long-poll test and I suspect these functions are the hot parts - adding a bunch of code to the inner loops seems like it may cause sadness

Reply:
Darn, good catch.

The existing code was structured so that the reference_matmul_element_<THETYPE> functions could just read LHS/RHS values straight out of buffers. Now it becomes `read_matrix_element` calls every time. This is indeed going to dominate profiles, i.e. make tests much slower (one could verify by looking at `ctest` latencies).  The solution (while retaining your PR's idea that the only thing we ever need is int32 reference matmul) could be either of the following:

* Either have a function to copy a generic matrix into a int32 matrix (rounding to nearest int32 / checking that values are exactly integral). Then call a `reference_matmul_element_i32` function like before.
* Or make this code a macro (closest C equivalent to a C++ template) so it can be type-specialized in a way that allows you to move the runtime type-dispatch out of that inner loop.

Reply:
Summarizing the discussion :- The matmul reference is generated using macro (closets to C++ template, it will be great to have googletest and C++ template for testing, hopefully in the future and we can discuss it). 

The reference should operate on the types of the underlying buffer types and it won't need any non-necessary casting. Thus, reference matmul will relatively be fast without any casting of if/else sequence required for the casting. We can further make the host reference faster by using simple blocking. 

For now, we are changing our initialization of buffers to use small integers whose ranges are controlled by the bitwidth of the datatype to make sure the accumulation on large k also stays in the range and a simple equality comparison can be performance to ensure correctness. The reference implementation should produce the correct result say our initialization change in the future.